src/transactions/TransactionFrame.cpp

```diff
@@ -740,6 +740,19 @@ TransactionFrame::applyOperations(SignatureChecker& signatureChecker,
                                   TransactionMeta& outerMeta)
 {
     ZoneScoped;
+    auto& internalErrorCounter = app.getMetrics().NewCounter(
+        {"ledger", "transaction", "internal-error"});
+
+    // We do not want to treat exception as internal-error for non-live ledger
+    // versions. The live ledger version is in the head of the ledger, and
+    // normally should be same as the app.getConfig().LEDGER_PROTOCOL_VERSION,
+    // which should also be same as maxAssertLedgerVersion here. However, right
+    // before a protocol upgrade, the LEDGER_PROTOCOL_VERSION (the software
+    // version) could lead the protocol version by 1, thus we need to offset it
+    // from the maxAssertLedgerVersion.
+    uint32_t maxAssertLedgerVersion =
+        app.getConfig().LEDGER_PROTOCOL_VERSION - 1;
+    uint32_t ledgerVersion = app.getConfig().LEDGER_PROTOCOL_VERSION;
```

Comment:
I see what you're saying, but shouldn't we always [initialize an variable ](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es20-always-initialize-an-object)to something?
In this case, would you suggest `0` being a better value to initialize to instead of `LEDGER_PROTOCOL_VERSION`?

Reply:
Yeah that's a good point. What about using https://en.cppreference.com/w/cpp/utility/optional? If `ledgerVersion` is accessed prior to it being set, an exception will be thrown, leading to a `txINTERNAL_ERROR`

Reply:
That should work, I've changed it to using `std::optional`